fix: fake native tokens need better UX error handling

salimtb opened the pull request:
## **Description**

In response to recent scam incidents we're adding mitigations to help prevent users from adding malicious networks and being shown inaccurate token values.

Implementation of this ticket will:

Show N/A for a native token value, when the native token symbol does not match the chain ID as publicly known
Show a tooltip that when tapped shows a bottom sheet explainer
The explainer directs the user to update their Network details for the network the token balance is requested for
User story

As a user I want to be informed that a token balance might be inaccurate, so that I don't make decisions to transfer or buy assets based on this information
As a user I want to be informed when a custom network might be trying to scam me


## **Related issues**

Fixes: [#1354](https://github.com/MetaMask/mobile-planning/issues/1354)

## **Manual testing steps**

1. Go to edit network page and choose Polygon
2.change the symbol from MATIC to ETH
3. Go to the wallet page

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

https://github.com/MetaMask/metamask-mobile/assets/26223211/af65f12b-0430-443a-9631-7b8beb188ee8


### **After**

https://drive.google.com/drive/folders/1lcKOyC7U1r9OW2qt7BjsReGdG-02EysJ?usp=sharing


<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

salimtb committed:
fix: fake native tokens need better UX error handling

salimtb committed:
fix: design review comments

salimtb commented:
figma : https://www.figma.com/file/T0z1ZJ81y8YlaKzoLGraTt/Verify-and-pre-populate-network-details-(mobile)?node-id=9%3A48702&mode=dev

Cal-L commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -111,6 +113,22 @@ const createStyles = (colors) =>
       color: colors.text.default,
     },
+    inputWithError: {
+      ...fontStyles.normal,
```
use typography from theme instead of fontStyles

Cal-L commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -111,6 +113,22 @@ const createStyles = (colors) =>
+      color: colors.text.default,
+    },
+    inputWithFocus: {
+      ...fontStyles.normal,
```
Same comment on typography

Cal-L commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -140,6 +161,12 @@ const createStyles = (colors) =>
+      paddingVertical: 8,
+      fontSize: 14,
+      color: colors.text.default,
+      ...fontStyles.normal,
```
Same comment for typography

Cal-L commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -358,6 +393,20 @@ class NetworkSettings extends PureComponent {
     } else {
       this.setState({ addMode: true });
     }
+
```
It's typically not good practice having data fetching logic within the component themselves. Could we abstract this into the `useIsOriginalNativeToken` hook instead?

Cal-L replied:
You can create a HOC wrapper that exposes the `useIsOriginalNativeToken` hook to this class component

salimtb replied:
done :)

Cal-L commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -358,6 +393,20 @@ class NetworkSettings extends PureComponent {
     }
+
+    axios
+      .get('https://chainid.network/chains.json')
```
Refer to abstracting this out into a constant

Cal-L commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -926,18 +1057,42 @@ class NetworkSettings extends PureComponent {
+                  {strings('wallet.chain_id_currently_used')}{' '}
+                  <Text
+                    style={styles.link}
+                    onPress={() => {
```
Could we abstract the inline logic into a function + add some comments to explain the functionality?

Cal-L commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -995,19 +1150,44 @@ class NetworkSettings extends PureComponent {
+                  {strings('wallet.use_the_currency_symbol')}{' '}
+                  <Text
+                    style={styles.link}
+                    onPress={() => {
```
Same comment regarding abstracting the function + add comments for functionality

Cal-L commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -91,6 +96,13 @@ import { selectDetectedTokens } from '../../../selectors/tokensController';
+  ButtonIconVariants,
+} from '../../../../app/component-library/components/Buttons/ButtonIcon';
+import Box from '../../UI/Ramp/components/Box';
+import ButtonSecondary from '../../../../app/component-library/components/Buttons/Button/variants/ButtonSecondary';
```
Replace ButtonSecondary with `Button` instead. You can then pass in the `variant` prop that indicates if it should be secondary. Ex.

`<Button variant={ButtonVariants.Secondary} />`

Cal-L commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -135,6 +155,66 @@ const Tokens: React.FC<TokensI> = ({ tokens }) => {
+        <ButtonIcon
+          iconName={IconName.Danger}
+          onPressIn={(e) => {
+            e.preventDefault();
```
I've seen this on web before but not so sure on mobile. What default behavior are you preventing here? Same question for stopPropogation. Could we also abstract this into a function + add comments

Cal-L commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -135,6 +155,66 @@ const Tokens: React.FC<TokensI> = ({ tokens }) => {
+          </Text>
+        </Box>
+        <Box style={styles.boxContent}>
+          <ButtonSecondary
```
Replace with Button and pass variant instead

Cal-L commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -135,6 +155,66 @@ const Tokens: React.FC<TokensI> = ({ tokens }) => {
+
+        <Box style={styles.boxContent}>
+          <Text>
+            {`This network doesn't match its associated chain ID or name. Many popular tokens use the name ${ticker}, making it a target for scams. Scammers may trick you into sending them more valuable currency in return. Verify everything before you continue.`}
```
Use translated string instead of hardcode

Cal-L commented on app/components/UI/Tokens/styles.ts:
```diff
@@ -100,6 +100,44 @@ const createStyles = (colors: Colors) =>
+    },
+
+    textMoadlHeader: {
+      ...fontStyles.bold,
```
Use typography from theme

Cal-L replied:
Looks like this didn't change?

Cal-L commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -229,12 +309,25 @@ const Tokens: React.FC<TokensI> = ({ tokens }) => {
     if (primaryCurrency === 'ETH') {
       mainBalance = balanceValueFormatted;
       secondaryBalance = balanceFiat;
+      if (asset.isETH) {
```
Let's add some comments to explain this logic

Cal-L commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -229,12 +309,25 @@ const Tokens: React.FC<TokensI> = ({ tokens }) => {
     } else {
       mainBalance = !balanceFiat ? balanceValueFormatted : balanceFiat;
       secondaryBalance = !balanceFiat ? balanceFiat : balanceValueFormatted;
+      if (asset.isETH) {
```
Let's add some comments to explain this logic and/or abstract into a function

Cal-L reviewed (commented):
Left some comments. Looks like there are failing unit tests as well

salimtb committed:
fix: fix units tests

salimtb committed:
fix: fix PR comments

salimtb committed:
fix: add HOC

salimtb committed:
fix: fix total balance

salimtb committed:
fix: fix form

salimtb committed:
fix: fix add custom network form bug

Cal-L commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -136,6 +155,68 @@ const Tokens: React.FC<TokensI> = ({ tokens }) => {
+    >
+      <Box style={styles.box}>
+        <View style={styles.notch} />
+        <SheetHeader title={'This is a potential scam'} />
```
Is there a translation string for this?

Cal-L commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -1219,6 +1420,11 @@ const mapStateToProps = (state) => ({
+  connect(mapStateToProps, mapDispatchToProps),
+  withIsOriginalNativeToken,
+)(NetworkSettings);
+// export default connect(mapStateToProps, mapDispatchToProps)(NetworkSettings);
```
Remove unused code

Cal-L reviewed (commented):
Left small comments

Cal-L reviewed (dismissed):
LGTM

cortisiko reviewed (changes requested):
See my comment above

salimtb committed:
fix: fix QA comments

salimtb commented:
> @salimtb We should double-check the logic when it comes to infura supported networks. I noticed that when on mainnet I see the warning next to ETH. See [here](https://app.screencast.com/TxY9jW3qOgcG4). [This](https://app.screencast.com/2vWR8lwArTVcQ) is what the network overview for `mainnet` shows

hello @cortisiko ,
thank you for your comment, the issue has now been resolved.

just to clarify a few things, at the bottom of the name you see a message that suggests using ETH as the name, this must be Ethereum mainnet, so thanks for reporting that, it's now fixed.

you shouldn't also see the red border in the filed name, so I've fixed that too.
the suggested network name came from [here](https://chainid.network/chains.json), you can also see the design [here](https://www.figma.com/file/T0z1ZJ81y8YlaKzoLGraTt/Verify-and-pre-populate-network-details-(mobile)?node-id=9%3A48702&mode=dev)

salimtb committed:
fix: fix linea testnet case

salimtb committed:
fix: fresh installed app bug

cortisiko reviewed (approved):
Looks good from a testing perspective

wachunei commented:
Hey @salimtb I just realized there were additions to the Ramp folder, to be exact the files:
- `app/components/UI/Ramp/hooks/useIsOriginalNativeTokenSymbol.test.ts`
- `app/components/UI/Ramp/hooks/useIsOriginalNativeTokenSymbol.ts`

These must have been added before our `CODEOWNERS` folders were listed so I was not aware of this change. I want to kindly ask for these two files to be moved to the corresponding code domain since these are not used within our feature.

cc: @Cal-L @cortisiko

salimtb commented:
> * app/components/UI/Ramp/hooks/useIsOriginalNativeTokenSymbol.test.ts

hey @wachunei ,
hooks now moved on this [PR](https://github.com/MetaMask/metamask-mobile/pull/9177) , can i have a review pls
